app/models/feed_event.rb

```diff
@@ -42,4 +47,30 @@ def self.record_journey_for(user, article:, category:)
         article: article,
       )
   end
+
+  private
+
+  def update_article_counters_and_scores
+    return unless article
+
+    impressions = article.feed_events.where(category: "impression")
+    return if impressions.empty? # This should not happen — but no need to continue if it does
+
+    clicks = article.feed_events.where(category: "click")
+    reactions = article.feed_events.where(category: "reaction")
+    comments = article.feed_events.where(category: "comment")
+
+    distinct_impressions_score = impressions.distinct.count(:user_id)
+    distinct_clicks_count = clicks.distinct.count(:user_id)
+    reactions_score = reactions.distinct.count(:user_id) * REACTION_SCORE_MULTIPLIER
+    comments_score = comments.distinct.count(:user_id) * COMMENT_SCORE_MULTIPLIER
+
+    score = (distinct_clicks_count + reactions_score + comments_score) / distinct_impressions_score.to_f
```

Comment:
So this is the real reason I mentioned rolling up events - less about saving raw storage space, and more that over time the _number_ of records has the potential to make these queries rather slow for what's supposed to be a fairly lightweight counter bump.

I think wrapping this in a `ThrottledCall` keyed per article ([as with billboard events](https://github.com/forem/forem/blob/c72c52dfdb3b2cea7ad69b11e904a8dbdef11879/app/controllers/billboard_events_controller.rb#L20-L32)) is fair for now, and we can revisit rolling up later

Reply:
Oh whoops I forgot to mention - I'm a little iffy about the `distinct` clauses here, since they would drop a _lot_ of information. For example no matter how many times a user comes across an article in their feed, only one impression of theirs would ever count - although maybe that's the intention, but then that would make collecting data about article position and such (or even ever creating the extra impressions in the first place) rather unnecessary

Reply:
This is the intention strictly for this score calculation, which is sort of a fusion number meant to represent total outcomes. Definitely agree with you on the minute data, where we are not rolling up in this way. The _counts_, and individual data are full captures where we can retain all the data, or as far as we can do so efficiently.